layer/training/runtime/model_trainer.py

```diff
@@ -38,7 +46,71 @@ class TrainContextDataclassMixin:
     train_index: Optional[str] = None
 
 
+cpu_used_temp = None  # pylint: disable=C0103
+start_time_temp = None  # pylint: disable=C0103
+step_value = 0  # pylint: disable=C0103
+
+
+def get_gpu_metrics(logger: Logger) -> Dict[str, Dict[str, float]]:
+    metrics = {}
+    try:
+        for gpu in nvsmi.get_gpus():
+            metrics[gpu.id] = {
+                "utilisation": gpu.gpu_util,
+                "mem_utilisation": round(gpu.mem_util, 2),
+                "mem_used": gpu.mem_used,
+                "mem_total": gpu.mem_total,
+                "id": gpu.id,
+                "name": gpu.name,
+            }
```

Comment:
nit: I'd put `nvsmi.get_gpus()` into a variable and only wrap that line with try..except